rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+
+- Attribute sets and lists could start on the same line as the if keywords, saving an indentation level on their body:
+  ```nix  
+  #1a
+  if builtins.length matches != 0 then {
+    inherit path matches;
+  } else if path == /. then [
+    1
+    2
+  ] else
+    go (dirOf path);
+  ```
+  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
+- Have the `then` on the start of the next line, directly followed by the if body:
+  ```nix
+  #1b
+  if builtins.length matches != 0
+    then { inherit path matches; }
+  else if path == /.
+    then [
+      1
+      2
+    ]
+  else go (dirOf path);
+
+  #1c
+  if builtins.length matches != 0
+  then { inherit path matches; }
+  else if path == /.
+  then [
+    1
+    2
+  ]
+  else go (dirOf path);
+  ```
+
+### with, assert
+
+**Description**
+
+- The body after the statement starts on a new line, without indentation.
+- There may be exceptions for common idioms, in which the body already starts on the same line.
+
+**Examples**
+
+```nix
+with pkgs;
+assert foo == bar; {
+  meta.maintainers =
+    with lib.maintainers; [
+      some
+      people
+    ];
+}
+```
+
+### let
+
+**Description**
+
+- Let bindings are *always* multiline.
+- The "let" part is indented one level, but not the "in" part.
+- Each item in the "let" part is indented and starts on its own line.
+  For more details, see the [binders section](#binders).
+- The "in" part starts on a new line.
+
+**Examples**
+
+```nix
+let
+  foo = "bar";
+in
+if foo == "bar" then
+  "hello"
+else
+  "world"
+```
+
+### Attribute sets and lists
+
+**Description**
+
+- Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - Empty lists and attribute sets are written as`[ ]` and`{ }`, respectively.
+- Lists and attribute sets with multiple items are liberally expanded.
+  - They can only be on a single line if:
+    - They contain at most one element
+    - Fit on the line
+  - As described under [Binders](#binders) below, nested attribute sets are always expanded.
+
+**Examples**
+
+```nix
+#1
+[
+  { }
+  { foo = "bar"; }
+  {
+    foo = {
+      bar = "baz";
+    };
+  }
+  { foo.bar = "baz"; }
+]
+
+#2
+[
+  [ 1 ]
+  [
+    2
+    3
+  ]
+]
+
+#3
+[
+  [
+    1
+    2
+    3
+  ]
+]
+
+#4
+[
+  {
+    mySingletons = [
+      [
+        ({
+          # stuff in there
+        })
+      ]
+    ];
+  
+    mySingletons' = [
+      [
+        (function call)
+      ]
+    ];
+  }
+]
+```
+
+**Drawbacks**
+
+- Singleton lists may use a lot of indentation
+
+**Rationale and alternatives**
+
+- Have a special compact form for singleton lists, to reduce the indentation level and remove two additional lines
+  ```nix
+  foo = [ {
+    # content
+  } ];
+  ```
+- Be a bit less eager about expanding lists and attribute sets (for example by allowing up to n elements on a single line).
+  ```nix
+  [ x y ]
+  ```
+
+### Binders and inherit
```

Comment:
This is a binder: https://duckduckgo.com/?t=ffab&q=binder&iax=images&ia=images

```suggestion
### Assignments and inherit
```

Reply:
First time I see _binders_.

```suggestion
### Bindings and inherit
```

Perhaps this is less specific (although I wouldn't know exactly what it refers to), but it does not create ambiguity.

Reply:
Changed it to bindings! https://github.com/nix-rfc-101/rfcs/commit/9ac50f80ab4aafa6fe773a54a8549e3e01ad7fec